### PR TITLE
Support testing with Postgres and MySQL for user management backend

### PR DESCRIPTION
**Pull request by ivov**

### CODE REVIEW

**Review by Susangriffis — APPROVED** [review]

File: packages/cli/test/integration/shared/testDb.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+				),
+			),
+		);
+	}
+
+	// MySQL truncation requires globals, which cannot be safely manipulated by parallel tests
+	if (dbType === 'mysqldb') {
+		await Promise.all(
+			entities.map(async (entity) => {
+				await Db.collections[entity]!.delete({});
```
Comment (BHesseldieck) [inline packages/cli/test/integration/shared/testDb.ts:142]: it shouldn't matter for tests, but do we also want to reset the auto-increment counter to stay consistent with the truncating dbs?

**Review by BHesseldieck — APPROVED** [review]
amazing what you have done here 🚀  laying an amazing foundation for tests in n8n 🎖️ 

just some questions/remarks/ideas

File: packages/cli/src/databases/entities/User.ts [hunk 2]
```diff
@@ -92,6 +92,15 @@ export class User {
...
 		nullable: true,
+		transformer: {
+			to: (answers: object) => answers,
+			from: (answers: object | string) => {
+				// Postgres stores and returns string
+				return typeof answers === 'string'
+					? (JSON.parse(answers) as IPersonalizationSurveyAnswers)
+					: answers;
+			},
+		},
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/databases/entities/User.ts]: Done. Tangent, but not sure I understand the logic for `idStringifier`. I would have kept unidirectional like this:

```ts
export const idStringifier = {
	to: (value: string): number => Number(value),
	from: (value: number): string => value.toString(),
};
```

File: packages/cli/src/databases/mysqldb/migrations/1636626154933-CreateUserManagement.ts [hunk 3]
```diff
@@ -9,261 +9,168 @@ export class CreateUserManagement1636626154933 implements MigrationInterface {
 	public async up(queryRunner: QueryRunner): Promise<void> {
 		const tablePrefix = config.get('database.tablePrefix');
 
+		await queryRunner.query('SET @@GLOBAL.FOREIGN_KEY_CHECKS=1;');
+		await queryRunner.query('SET @@SESSION.FOREIGN_KEY_CHECKS=1;');
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/src/databases/mysqldb/migrations/1636626154933-CreateUserManagement.ts]: Leftover from debugging, sorry.

File: packages/cli/test/teardown.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+	if (dbType === 'postgresdb') {
+		const bootstrap = await createConnection(getBootstrapPostgresOptions());
+
+		const results: { db_name: string }[] = await bootstrap.query(
+			'SELECT datname as db_name FROM pg_database;',
+		);
+
+		const promises = results
+			.filter(({ db_name }) => db_name.startsWith('n8n_test_pg_'))
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/test/teardown.ts]: Turns out Postgres DB names [cannot start with a number](https://www.postgresql.org/docs/7.0/syntax525.htm ).

So went with `pg_${Date.now()}_n8n_test` and analogous for MySQL, checking on both ends.

File: packages/cli/test/teardown.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+		bootstrap.close();
+	}
+
+	if (dbType === 'mysqldb') {
+		const user = config.get('database.mysqldb.user');
+		const password = config.get('database.mysqldb.password');
+		const host = config.get('database.mysqldb.host');
+
+		exec(`echo "DROP DATABASE n8n_bs_mysql" | mysql -h ${host} -u ${user} -p${password}`);
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/cli/test/teardown.ts]: MySQL database dropping happens here at `testDb.terminate()` in `afterAll()`

https://github.com/n8n-io/n8n/blob/1a6a563bcca2883204498b9b3b16d36ec1cce5cc/packages/cli/test/integration/shared/testDb.ts#L94 

What you prolly mean is the no-leftover check (as in PG) to ensure no MySQL test DBs remain from test run interruption? If so, added just now.

File: packages/cli/test/integration/shared/testDb.ts (hunk 1 above)
Comment (ivov) [reply packages/cli/test/integration/shared/testDb.ts:141]: Done.

File: packages/cli/src/databases/entities/User.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (BHesseldieck) [reply packages/cli/src/databases/entities/User.ts]: I had it like this, but for whatever reason it was failing on fetching workflows, so I had to handle `null | undefined`

**Review by BHesseldieck — APPROVED** [review]
amazing 🚀

**Pull request merged**